common/src/java/org/apache/hive/http/HttpServer.java

```diff
@@ -604,108 +715,65 @@ void setContextAttributes(Context ctx, Map<String, Object> contextAttrs) {
 
   private void createWebServer(final Builder b) throws IOException {
     // Create the thread pool for the web server to handle HTTP requests
-    QueuedThreadPool threadPool = new QueuedThreadPool();
+    threadPool = new QueuedThreadPool();
     if (b.maxThreads > 0) {
       threadPool.setMaxThreads(b.maxThreads);
     }
     threadPool.setDaemon(true);
     threadPool.setName(b.name + "-web");
 
     this.webServer = new Server(threadPool);
-    this.appDir = getWebAppsPath(b.name);
-    this.webAppContext = createWebAppContext(b);
-
-    if (b.useSPNEGO) {
-      // Secure the web server with kerberos
-      setupSpnegoFilter(b, webAppContext);
-    }
-
-    if (b.enableCORS) {
-      setupCORSFilter(b);
-    }
-
-    Map<String, String> xFrameParams = setHeaders();
-    if (b.xFrameEnabled) {
-      setupXframeFilter(b,xFrameParams);
-    }
-
-    if (b.disableDirListing) {
-      disableDirectoryListingOnServlet(webAppContext);
-    }
+    this.webAppContexts = new LinkedList<>();
 
-    initializeWebServer(b, threadPool.getMaxThreads());
+    initializeWebServer(b);
   }
 
-  private void initializeWebServer(final Builder b, int queueSize) throws IOException {
+  private void initializeWebServer(final Builder b) throws IOException {
     // Set handling for low resource conditions.
     final LowResourceMonitor low = new LowResourceMonitor(webServer);
     low.setLowResourcesIdleTimeout(10000);
     webServer.addBean(low);
 
-    Connector connector = createChannelConnector(queueSize, b);
-    webServer.addConnector(connector);
-
-    RewriteHandler rwHandler = new RewriteHandler();
-    rwHandler.setRewriteRequestURI(true);
-    rwHandler.setRewritePathInfo(false);
-
-    RewriteRegexRule rootRule = new RewriteRegexRule();
-    rootRule.setRegex("^/$");
-    rootRule.setReplacement(b.contextRootRewriteTarget);
-    rootRule.setTerminating(true);
-
-    rwHandler.addRule(rootRule);
-    rwHandler.setHandler(webAppContext);
-
-    // Configure web application contexts for the web server
-    ContextHandlerCollection contexts = new ContextHandlerCollection();
-    contexts.addHandler(rwHandler);
-    webServer.setHandler(contexts);
+    // Configure the global context handler
+    portHandlerWrapper = new PortHandlerWrapper();
+    webServer.setHandler(portHandlerWrapper);
 
+    // Configure the web server connector and port handler to listen on
+    ContextHandlerCollection portHandler = addWebApp(b);
```

Comment:
should we do this closer to actual ref usage?

Reply:
I could move it only a few lines below and added a comment:

```
    // Configures the web server connector and port handler to listen on
    // Also creates and adds the web application context to the server to which the servlets will be added
    ContextHandlerCollection portHandler = createAndAddWebApp(b);

    addServlet("jmx", "/jmx", JMXJsonServlet.class);
    addServlet("conf", "/conf", ConfServlet.class);
    addServlet("stacks", "/stacks", StackServlet.class);
    addServlet("conflog", "/conflog", Log4j2ConfiguratorServlet.class);
```